Add a window switcher widget

Kasper24 opened the pull request:
This is based on elenapan window_switcher widget, modified it to add a thumbnail preview and to work nicely with bling. Let me know any further changes I need to make

JavaCafe01 commented on widget/window_switcher.lua:
```diff
@@ -0,0 +1,400 @@
+    end
+    local client_height = opts.client_height or dpi(250)
+    local client_margin = opts.client_margin or dpi(10)
+    local background = opts.background or "#00000000"
```
```suggestion
    local background = beautiful.window_switcher_widget_bg or "#000000"
```

I would have the default bg color as black, instead of transparent just because all the other default bgs for widgets are back and because not all of our users are probably using compositors.

Also, this should be a theme var instead of an opt.

JavaCafe01 commented on widget/window_switcher.lua:
```diff
@@ -0,0 +1,400 @@
+    local client_height = opts.client_height or dpi(250)
+    local client_margin = opts.client_margin or dpi(10)
+    local background = opts.background or "#00000000"
+    local border_radius = opts.border_radius or dpi(0)
```
```suggestion
    local border_radius = beautiful.window_switcher_widget_border_radius or dpi(0)
```

JavaCafe01 reviewed (changes requested):
Just tried out your widget and it does work and looks amazing. Just some minor things that I want you to change (if you want different beautiful theme var names, go ahead). After these fixes are done, you'll have to add this to the docs (or I can, if you don't want to) and we can merge!

JavaCafe01 commented:
Just to note down -

```lua
bling.widget.window_switcher.enable {}

-- Somewhere in ur keys ->

        awful.key({modkey}, "Tab", function()
            -- awful.client.focus.history.previous()
            awesome.emit_signal("bling::window_switcher::visibility",
                                mouse.screen)
            -- if client.focus then client.focus:raise() end
        end, {description = "Window Switcher", group = "client"})
```

JavaCafe01 commented:
@Kasper24 Also add yourself to [AUTHORS.md](https://github.com/BlingCorp/bling/blob/master/AUTHORS.md when your done.

Nooo37 commented on widget/window_switcher.lua:
```diff
@@ -0,0 +1,400 @@
+                {
+                    widget = wibox.container.background,
+                    id = "bg_role",
+                    forced_height = item_height,
```
`item_height` isn't declared

Nooo37 commented on widget/window_switcher.lua:
```diff
@@ -0,0 +1,400 @@
+
+    -- Minimize originally minimized clients
+    local s = awful.screen.focused()
+    local clients = s.selected_tag:clients()
```
`clients` is unused anywhere else in the block?

Piterden commented on widget/window_switcher.lua:
```diff
@@ -0,0 +1,400 @@
+        if (font_icons) ~= nil then
+            return {
+                widget = wibox.widget.textbox,
+                id     = 'text_icon',
+                font   = font_icons_font,
+                forced_width = dpi(50),
+                align  = "center",
```
Mixing single and double quotes is a kind of bad manners...

JavaCafe01 replied:
It is, but it really depends on the repository you are commiting too, and in this case I don't really care about it... Lol

Piterden commented on widget/window_switcher.lua:
```diff
@@ -0,0 +1,400 @@
+                align  = "center",
+                valign = "center",
+            }
+        else
```
Why do you need else if you use return?
```lua
if cond then
    return something
end
return anotherthing
```

Piterden commented on widget/window_switcher.lua:
```diff
@@ -0,0 +1,400 @@
+
+    local type = opts.type or "thumbnail"
+    local client_width
+    if type == "thumbnail" then
+        client_width = opts.client_width or dpi(150)
+    else
+        client_width = opts.client_width or dpi(500)
+    end
```
Why not like below?
```lua
local client_width = opts.client_width or dpi(type == "thumbnail" and 150 or 500)
```

JavaCafe01 replied:
one can argue that the code above yours is readable :smile:

Kasper24 commented:
I fixed all the issues you guys mentioned, refactored, reformatted, improved and add a lot more styling options. I'm not yet sure which options should belong to beautiful and which not, maybe @JavaCafe01 can help me with that?

JavaCafe01 commented:
@Kasper24 I think this PR is still in hold because it doesn't work on Awesome stable

Kasper24 commented:
@JavaCafe01 Fixed in the a83f7fc6c9c169754ebefefd44f88a6c66324d4f

It will look a little worse on stable, as vertical/horizontal_fit_policy didn't exist back then so the thumbnail wouldn't scale to fit correctly. There might be another way of doing this, but I'm not sure

JavaCafe01 commented:
@Kasper24 Before I do a complete review, just a few things. I liked how things were without the scroll widget - the text was wrapped and centered (the scroll widget also has an impact on performance, doesn't start in the center, and has some bugs). Maybe later on we can add a `widget_structure` thing like how @undefinedDarkness did it to allow for custom structures. Also, it would be nice if the client icon showed on the fallback as well (if it is supposed to, then this is a scroll widget bug).

Nooo37 commented on widget/window_switcher.lua:
```diff
@@ -0,0 +1,390 @@
+                widget = wibox.container.margin
+            },
+            border_width = border_width,
+            border_color = border_color,
```
Do `shape_border_color` and `shape_border_width` here instead for stable compatibility

Nooo37 reviewed (commented):
The titles won't show up on stable anymore. It worked yesterday even with the scroll container. Don't exactly know what changed. 

On the git version, the previews are all squashed into the same ratio which distortes the content while they keep their ratio on the stable version. I don't know what the expected behavior here would be but I would be ok with both.

~~(Sidenote:) I'm also experiencing that I can't close the popup completely after pressing my hide-key (ie the keygrabber will stop but the widget won't hide, I think it gets toggled on and off real quick). Maybe someone can reproduce but might as well just be a me-issue. Just let me know. I have the widget initalized with `bling.widget.window_switcher.enable { hide_window_switcher_key = 'e' }` and starting it with `awesome-client "awesome.emit_signal('bling::window_switcher::visibility', mouse.screen)"`. Would be cool to know if I have to look into my own config more ;-;~~ 
nvm I had the module enabled twice in my config and that seems to be the sideeffect :facepalm:

Kasper24 commented:
@Nooo37 
> The titles won't show up on stable anymore. It worked yesterday even with the scroll container. Don't exactly know what changed.

I honestly don't know? I did make the name margins customizable and changed the default margins for the thumbnail type widget, but I don't see why if it breaks it would only break on stable?

> On the git version, the previews are all squashed into the same ratio which distortes the content while they keep their ratio on the stable version. I don't know what the expected behavior here would be but I would be ok with both.

I commented on this [here](https://github.com/BlingCorp/bling/pull/84#issuecomment-899023983). I'll make this into an option, but is there a way I could also achieve the on stable?

gennarocc commented:
Current iteration seems to work fine for me.

Nooo37 commented on widget/window_switcher.lua:
```diff
@@ -0,0 +1,350 @@
+                                id = "text_role",
+                                widget = wibox.widget.textbox
+                            },
+                            margins = name_margins,
```
I was mentioning that the text started to disappear on stable a couple weeks ago (sorry for that :sweat_smile:) and it's because of the margins I think: If I comment that line out, the text does appear on stable. Maybe just do that as a quick fix? I failed to see why that is but even if I set all margins to 0 in the `name_margins` initalization then it still doesn't appear.

Nooo37 reviewed (commented):
Is merge ready once the titles appear on 4.3 (see the comment for more info). Do you want to do the docs for that widget or should we do them?

Kasper24 commented:
@Nooo37 Fixed at 1a94b02e2a8498c12a384fd9ef86c12cc0c29423
